internal/uvm/start.go

```diff
@@ -1,57 +1,93 @@
+type gcsLogEntryStandard struct {
+	Time    time.Time    `json:"time"`
+	Level   logrus.Level `json:"level"`
+	Message string       `json:"msg"`
+}
+
+type gcsLogEntry struct {
+	gcsLogEntryStandard
+	Fields map[string]interface{}
+}
 
-func parseLogrus(r io.Reader) {
-	j := json.NewDecoder(r)
-	logger := logrus.StandardLogger()
-	for {
-		e := logrus.Entry{Logger: logger}
-		err := j.Decode(&e.Data)
-		if err == io.EOF || err == _ERROR_CONNECTION_ABORTED {
-			break
+// FUTURE-jstarks: Change the GCS log format to include type information
+//                 (e.g. by using a different encoding such as protobuf).
+func (e *gcsLogEntry) UnmarshalJSON(b []byte) error {
+	// Default the log level to info.
+	e.Level = logrus.InfoLevel
+	if err := json.Unmarshal(b, &e.gcsLogEntryStandard); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(b, &e.Fields); err != nil {
+		return err
+	}
+	// Do not allow fatal or panic level errors to propagate.
+	if e.Level < logrus.ErrorLevel {
+		e.Level = logrus.ErrorLevel
+	}
+	// Clear special fields.
+	delete(e.Fields, "time")
+	delete(e.Fields, "level")
+	delete(e.Fields, "msg")
+	// Normalize floats to integers.
+	for k, v := range e.Fields {
+		if d, ok := v.(float64); ok && float64(int64(d)) == d {
+			e.Fields[k] = int64(d)
 		}
-		if err != nil {
-			// Something went wrong. Read the rest of the data as a single
-			// string and log it at once -- it's probably a GCS panic stack.
-			logrus.Error("gcs log read: ", err)
-			rest, _ := ioutil.ReadAll(io.MultiReader(j.Buffered(), r))
-			if len(rest) != 0 {
-				logrus.Error("gcs stderr: ", string(rest))
+	}
+	return nil
+}
+
+func parseLogrus(vmid string) func(r io.Reader) {
+	return func(r io.Reader) {
+		j := json.NewDecoder(r)
+		e := logrus.NewEntry(logrus.StandardLogger())
+		fields := e.Data
+		for {
+			for k := range fields {
+				delete(fields, k)
 			}
-			break
-		}
-		msg := e.Data["msg"]
-		delete(e.Data, "msg")
-		lvl := e.Data["level"]
-		delete(e.Data, "level")
-		e.Data["vm.time"] = e.Data["time"]
-		delete(e.Data, "time")
-		switch lvl {
-		case "debug":
-			e.Debug(msg)
-		case "info":
-			e.Info(msg)
-		case "warning":
-			e.Warning(msg)
-		case "error", "fatal":
-			e.Error(msg)
-		default:
-			e.Info(msg)
+			gcsEntry := gcsLogEntry{Fields: e.Data}
+			err := j.Decode(&gcsEntry)
+			if err != nil {
+				// Something went wrong. Read the rest of the data as a single
+				// string and log it at once -- it's probably a GCS panic stack.
+				if err != io.EOF && err != _ERROR_CONNECTION_ABORTED && err != syscall.WSAECONNRESET {
+					logrus.WithFields(logrus.Fields{
+						logfields.UVMID: vmid,
+						logrus.ErrorKey: err,
+					}).Error("gcs log read")
+				}
+				rest, _ := ioutil.ReadAll(io.MultiReader(j.Buffered(), r))
+				rest = bytes.TrimSpace(rest)
+				if len(rest) != 0 {
+					logrus.WithFields(logrus.Fields{
+						logfields.UVMID: vmid,
+						"stderr":        string(rest),
+					}).Error("gcs terminated")
```

Comment:
We should only log at Error level if GCS terminates abnormally or we hit an error when reading.

Reply:
On second thought, maybe if the connection closes gracefully then `len(rest)` would be 0, and we wouldn't log the error here?

Reply:
Right. GCS should not be writing non-JSON, non-whitespace output to stderr unless something has gone wrong.

Maybe I should change "gcs terminated" to "gcs panic", since that's the only expected case where this happens. Although technically it could happen if the GCS is killed (or the VM is terminated) half-way through writing some JSON entry.

Reply:
I think it's okay the way it is, it just took me a minute to understand it.